File: NeetCode/03-LinkedLists/C++/DoublyLinkedList.cpp
//
// Created by Jerry on 8/1/2025.
//
#include <iostream>
using std::cout;
using std::endl;

class ListNode
{
public:
    int val_;
    ListNode *next = nullptr;
    ListNode *prev = nullptr;
    ListNode(int val)
    {
        val_ = val;
    }
};

class LinkedList
{
public:
    ListNode *head;
    ListNode *tail;

    LinkedList(int val)
    {
        head = new ListNode(-1);
        tail = new ListNode(-1);
        head->next = tail;
        tail->prev = head;
    }

    void insertFront(int val)
    {
        ListNode *newNode = new ListNode(val);
        newNode->next = head->next;
        newNode->prev = head;

        head->next->prev = newNode;
        head->next = newNode;
    }

    void insertEnd(int val)
    {
        ListNode* newNode = new ListNode(val);
        newNode->prev = tail->prev;
        newNode->next = tail;

        tail->prev->next = newNode;
        tail->prev = newNode; 
    }

    void removeFront()
    {
        head->next->next->prev = head;
        head->next = head->next->next;
    }

    void removeEnd()
    {
        tail->prev->prev->next = tail;
        tail->prev = tail->prev->prev;
    }

    void print()
    {
        ListNode *curr = head;
        while (curr != tail)
        {
            cout << curr->val_ << " ";
            curr = curr->next;
        }
        cout << endl;
    }
};

File: NeetCode/03-LinkedLists/Python/DoublyLinkedList.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None

# Implementation for Doubly Linked List
class LinkedList:
    def __init__(self, val):
        self.head = ListNode(-1)
        self.tail = ListNode(-1)
        self.head.next = self.tail
        self.tail.prev = self.head
    
    def insertFront(self, val):
        newNode = ListNode(val)
        newNode.prev = self.head
        newNode.next = self.head.next
        
        self.head.next.prev = newNode
        self.head.next = newNode
            
    def insertEnd (self, val):
        newNode = ListNode(val)
        newNode.next = self.tail
        newNode.prev = self.tail.prev
        
        self.tail.prev.next = newNode
        self.tail.prev = newNode
        
    def removeFront(self):
        self.head.next.next.prev = self.head
        self.head.next = self.head.next.next
    
    def removeEnd(self):
        self.tail.prev.prev.next = self.tail
        self.tail.prev = self.tail.prev.prev  
    
    def print(self):
        curr = self.head.next
        
        while curr != self.tail:
            print(f"{curr.val} ")
            curr = curr.next
        print("\n")